### PR TITLE
[v4.1] Raven db 10535

### PR DESCRIPTION
**Pull request by maximburyak**

### CODE REVIEW

File: src/Raven.Client/Json/BlittableJsonReader.cs [hunk 1]
```diff
@@ -152,11 +155,32 @@ private bool SetToken(BlittableJsonToken token, object value)
...
                 case BlittableJsonToken.LazyNumber:
-                    SetToken(JsonToken.Float, (double)(LazyNumberValue)value);
+                    if (_readAsLazyNumber)
+                    {
+                        SetToken(JsonToken.Float, value);
+                    }
+                    else
+                    {
+                        LazyNumberValue lnv = (LazyNumberValue)value;
+                        if (lnv.TryParseUlong(out var ulongValue))
```
Comment (ayende) [inline src/Raven.Client/Json/BlittableJsonReader.cs]: `ULong`

File: src/Raven.Server/Documents/Patch/BlittableObjectInstance.cs [hunk 2]
```diff
@@ -1,58 +1,89 @@
...
         public readonly BlittableJsonReaderObject Blittable;
         public readonly string DocumentId;
         public HashSet<string> Deletes;
-        public Dictionary<string, BlittableObjectProperty> OwnValues = new Dictionary<string, BlittableObjectProperty>();
+        public Dictionary<string, BlittableObjectProperty > OwnValues = 
+            new Dictionary<string, BlittableObjectProperty>();
         public Dictionary<string, BlittableJsonToken> OriginalPropertiesTypes;
         public Lucene.Net.Documents.Document LuceneDocument;
         public IState LuceneState;
+        public List<AllocatedMemoryData> AllocatedMemory;
```
Comment (ayende) [inline src/Raven.Server/Documents/Patch/BlittableObjectInstance.cs]: This shouldn't exists

File: src/Raven.Server/Documents/Patch/BlittableObjectInstance.cs [hunk 3]
```diff
@@ -175,6 +223,11 @@ private JsValue TranslateToJs(BlittableObjectInstance owner, string key, Blittab
             DateTime? lastModified, string changeVector = null) : base(engine)
         {
             _parent = parent;
+            _root = parent;
+
+            while (_root != null && _root._parent != null)
```
Comment (ayende) [inline src/Raven.Server/Documents/Patch/BlittableObjectInstance.cs]: remove

File: src/Raven.Server/Documents/Patch/BlittableObjectInstance.cs [hunk 4]
```diff
@@ -228,5 +287,17 @@ private void RecordNumericFieldType(string key, BlittableJsonToken type)
                 OriginalPropertiesTypes = new Dictionary<string, BlittableJsonToken>();
             OriginalPropertiesTypes[key] = type;
         }
+
+        public void Dispose()
```
Comment (ayende) [inline src/Raven.Server/Documents/Patch/BlittableObjectInstance.cs]: remove

File: src/Raven.Server/Documents/Patch/ScriptRunner.cs [hunk 5]
```diff
@@ -549,6 +554,89 @@ private JsValue Regex(JsValue self, JsValue[] args)
...
+
+
+                var ret = func.Body.Body.As<ReturnStatement[]>();
+
+                if (ret.Length != 1)
+                {
+                    throw new InvalidOperationException("scalarToRawString(document, lambdaToField) lambda to field must contain a single return expression");
+                }
+                StaticMemberExpression staticMemberExpression = ret[0].Argument.As<StaticMemberExpression>();
+                var prop = staticMemberExpression.Property;
```
Comment (ayende) [inline src/Raven.Server/Documents/Patch/ScriptRunner.cs]: Need to validate that this holds

File: src/Raven.Server/Documents/Patch/ScriptRunner.cs [hunk 6]
```diff
@@ -549,6 +554,89 @@ private JsValue Regex(JsValue self, JsValue[] args)
...
+                }
+
+                JsValue secondParam = args[1];
+                if (args[1].IsObject() == false || secondParam.AsObject() is FunctionInstance == false)
+                    throw new InvalidOperationException("scalarToRawString(document, lambdaToField) must be called with a second lambda argument");
+
+
+
+                BlittableObjectInstance selfInstance = firstParam.AsObject() as BlittableObjectInstance;
+                FunctionInstance lambda = secondParam.AsObject() as FunctionInstance;
```
Comment (ayende) [inline src/Raven.Server/Documents/Patch/ScriptRunner.cs]: Replace all this with TryGetField from 4.1 branch

Comment (ayende) [reply src/Raven.Server/Documents/Patch/ScriptRunner.cs]: See: https://github.com/ravendb/ravendb/blob/v4.1/src/Raven.Server/Documents/Indexes/Static/JavaScriptIndex.cs#L479-L492

File: src/Raven.Server/Documents/Patch/ScriptRunner.cs [hunk 7]
```diff
@@ -608,13 +696,23 @@ public ScriptRunnerResult Run(JsonOperationContext jsonCtx, DocumentsOperationCo
...
                 finally
-                {
+                {                    
                     _refResolver.ExplodeArgsOn(null, null);
                     _docsCtx = null;
                     _jsonCtx = null;
                 }
             }
 
+            internal void ReleaseAllocations()
```
Comment (ayende) [inline src/Raven.Server/Documents/Patch/ScriptRunner.cs]: don't need this

File: src/Raven.Server/Documents/Patch/ScriptRunner.cs [hunk 8]
```diff
@@ -608,13 +696,23 @@ public ScriptRunnerResult Run(JsonOperationContext jsonCtx, DocumentsOperationCo
...
+            {
+                for (var i=0; i< _objectInstances.Count; i++)
+                {
+                    _objectInstances[i].Dispose();
+                }
+            }
+
+            private List<BlittableObjectInstance> _objectInstances = new List<BlittableObjectInstance>();
```
Comment (ayende) [inline src/Raven.Server/Documents/Patch/ScriptRunner.cs]: remove

File: src/Raven.Server/Documents/Patch/ScriptRunner.cs [hunk 9]
```diff
@@ -673,23 +771,39 @@ BlittableJsonReaderObject Clone(BlittableJsonReaderObject origin)
...
                         LuceneState = t.Item3
                     };
+                    _objectInstances.Add(blittableObjectInstance);
+                    return blittableObjectInstance;
                 }
                 if (o is Document doc)
                 {
-                    return new BlittableObjectInstance(engine, null, Clone(doc.Data), doc.Id, doc.LastModified);
+                    blittableObjectInstance = new BlittableObjectInstance(engine, null, Clone(doc.Data), doc.Id, doc.LastModified);
+                    _objectInstances.Add(blittableObjectInstance);
```
Comment (ayende) [inline src/Raven.Server/Documents/Patch/ScriptRunner.cs]: remove

File: src/Raven.Server/Documents/Patch/ScriptRunnerResult.cs [hunk 10]
```diff
@@ -51,6 +51,7 @@ public BlittableJsonReaderObject TranslateToObject(JsonOperationContext context,
 
         public void Dispose()
         {
+            _parent?.ReleaseAllocations();
```
Comment (ayende) [inline src/Raven.Server/Documents/Patch/ScriptRunnerResult.cs]: remove

File: src/Sparrow/Json/LazyStringValue.cs [hunk 11]
```diff
@@ -98,6 +98,8 @@ public int Length
         public AllocatedMemoryData AllocatedMemoryData;
         public int? LastFoundAt;
 
+        public bool HasMaterializedStringValue => _string != null;
```
Comment (ayende) [inline src/Sparrow/Json/LazyStringValue.cs]: Can remove

File: src/Sparrow/Json/Parsing/UnmanagedJsonParser.cs [hunk 12]
```diff
@@ -883,37 +882,19 @@ private void WriteUnicodeCharacterToStringBuffer(int val)
...
-                {
-                    pChars[i] = ' ';
-                }
-            }
             try
             {
-                // ReSharper disable once ReturnValueOfPureMethodIsNotUsed
-                double.Parse(_doubleStringBuffer, NumberStyles.Any, CultureInfo.InvariantCulture);
+                int numLength = _unmanagedWriteBuffer.SizeInBytes;
+                var tmpBuff = stackalloc byte[numLength];
```
Comment (ayende) [inline src/Sparrow/Json/Parsing/UnmanagedJsonParser.cs]: Let's avoid smashing the stack

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Client/Json/BlittableJsonReader.cs (hunk 1 above)
Comment (maximburyak) [reply src/Raven.Client/Json/BlittableJsonReader.cs]: OK

File: src/Raven.Server/Documents/Patch/BlittableObjectInstance.cs (hunk 2 above)
Comment (maximburyak) [reply src/Raven.Server/Documents/Patch/BlittableObjectInstance.cs]: OK

File: src/Raven.Server/Documents/Patch/BlittableObjectInstance.cs (hunk 3 above)
Comment (maximburyak) [reply src/Raven.Server/Documents/Patch/BlittableObjectInstance.cs]: OK

File: src/Raven.Server/Documents/Patch/BlittableObjectInstance.cs (hunk 4 above)
Comment (maximburyak) [reply src/Raven.Server/Documents/Patch/BlittableObjectInstance.cs]: OK

File: src/Raven.Server/Documents/Patch/ScriptRunner.cs (hunk 8 above)
Comment (maximburyak) [reply src/Raven.Server/Documents/Patch/ScriptRunner.cs]: OK

File: src/Raven.Server/Documents/Patch/ScriptRunner.cs (hunk 9 above)
Comment (maximburyak) [reply src/Raven.Server/Documents/Patch/ScriptRunner.cs]: OK

File: src/Raven.Server/Documents/Patch/ScriptRunnerResult.cs (hunk 10 above)
Comment (maximburyak) [reply src/Raven.Server/Documents/Patch/ScriptRunnerResult.cs]: OK

File: src/Sparrow/Json/LazyStringValue.cs (hunk 11 above)
Comment (maximburyak) [reply src/Sparrow/Json/LazyStringValue.cs]: OK

File: src/Sparrow/Json/Parsing/UnmanagedJsonParser.cs (hunk 12 above)
Comment (maximburyak) [reply src/Sparrow/Json/Parsing/UnmanagedJsonParser.cs]: OK

File: src/Raven.Server/Documents/Patch/ScriptRunner.cs (hunk 7 above)
Comment (maximburyak) [reply src/Raven.Server/Documents/Patch/ScriptRunner.cs]: OK

File: src/Raven.Server/Documents/Patch/ScriptRunner.cs (hunk 5 above)
Comment (maximburyak) [reply src/Raven.Server/Documents/Patch/ScriptRunner.cs]: ok

File: src/Raven.Server/Documents/Patch/ScriptRunner.cs (hunk 6 above)
Comment (maximburyak) [reply src/Raven.Server/Documents/Patch/ScriptRunner.cs]: ok

**Pull request merged**